### PR TITLE
Add MergeDataSegments pass

### PR DESCRIPTION
**Pull request by LegionMammal978**
Recently, I was writing a WASM module by hand, and I used a number of individual small data segments to store strings. I tried seeing if wasm-opt could combine these adjacent data segments, but found that it did not have any such pass. Thus, I've implemented a new MergeDataSegments pass to merge active data segments that are overlapping, adjacent, or near-adjacent, in order to save the space of storing multiple data-segment headers, and the time of processing them during instantiation.

It is designed to be as aggressive as possible, fully supporting multiple memories and accounting for non-constant-offset segments. Meanwhile, unless TNH is enabled, it is also designed to carefully replicate the original module's behavior w.r.t. out-of-bounds traps during instantiation: the goal is that there should be no observable difference in the output, short of unreliable tricks like reading a partially-instantiated `SharedArrayBuffer`.

In principle, this functionality might have been included in the existing MemoryPacking pass, but I believe that it makes sense to separate the primary functionalities of splitting vs. merging data segments, which require different forms of tracking. In that sense, I see MergeDataSegments as complementing the MemoryPacking pass. For instance, MemoryPacking requires that its input has no overlapping data segments, a property that MergeDataSegments is often able to ensure in its output.

Some implementation notes:

- Following the behavior of MemoryPacking, it reads TNH as asserting traps never happen during instantiation, so in particular data segments are never out of bounds. Also, it only considers TNH and not `--ignore-implicit-traps`, following MemoryPacking.
- The pass detects certain cases when a data segment is necessarily out-of-bounds: in that case, it simply emits the offending data segment last, and drops all remaining data segments. In principle, since the module cannot be fully instantiated, it could be even more aggressive with replacing every function body with `unreachable`, etc., but this is a small edge case.
- Following MemoryPacking, this pass assumes that unless a memory is imported, it is zero-initialized, and its initial and maximum sizes are exactly as declared. It also assumes that during data initialization, the memory cannot be modified by anything other than the declared data segments. It is my understanding that these assumptions are not affected by open-world vs. closed-world.
- The pass tries to be careful about integer limits, with only one small edge case remaining: As far as I can read the WASM spec, it permits a 64-bit memory to be exactly 2^64 bytes long, yet the pass only handles memories up to 2^64-1 bytes long. But in general, binaryen doesn't seem to be designed in a way that would allow that last byte.
- Overall, the pass modifies the module by rewriting its data segments, then modifying data-segment indices in function bodies. In the latter case, I assume that `ReFinalize()` is not needed, since the instructions and stack arguments are unchanged, only their indices are modified. Similarly, I mark `requiresNonNullableLocalFixups()` as `false`. In principle, the behavior of the memory instructions on active data segments can be simplified, but I figure that it's better to leave that to the more extensive modifications of MemoryPacking.
- The active-segment `threshold` of MemoryPacking does not always match the size heuristic of MergeDataSegments, so if run in alternation in certain cases, they could fight over splitting vs. merging the same two segments.
- Obviously, this pass has a lot of edge cases that need testing, but I'm not sure where the tests should be placed (`test/passes/`? `test/lit/passes/`?), nor how exactly they are formatted. I'm especially unsure how to test behavior around the `MAX_SEG_SIZE`.

### CODE REVIEW

**Comment by MaxGraey** [comment]
It looks like all of this was created using an LLM?  You should add tests to demonstrate how this data segments merges + check edge cases using lit tests (test/lit/passes/`<some-pass-name>`.wast). You also need to run fuzz tests for at least 5-7 hours.

**Comment by LegionMammal978** [comment]
No, I wrote everything in this PR myself, I'm not a big fan of LLMs' coding style. (Indeed, I asked an LLM to review my merge functions, and it kept wanting to add all sorts of extraneous steps.)

I think I see how to write the lit tests (create the file with the inputs, run `scripts/update_lit_tests.py`, and double-check that the outputs are correct). But how do I run the fuzz tests?

**Comment by MaxGraey** [comment]
> But how do I run the fuzz tests?

run from root dir:
```bash
./scripts/fuzz_opt.py
```

**Comment by LegionMammal978** [comment]
I've added tests for most of the functionality, and ironed out a few bugs. I've tried running `fuzz_opt.py`, but it keeps hitting errors not involving the new pass. From what I can tell, it's seemingly trying to use a `(pagesize 1)` memory from test cases in my `test/lit/passes/merge-data-segments.wast`, but V8 has no support for the custom page sizes proposal, so the fuzzer gets mad about it.

**Comment by kripken** [comment]
>  V8 has no support for the custom page sizes proposal

You may need a recent build of V8, yes. Another option is to manually disable that feature (iirc there is a set of such features at the top of the python script).

File: src/passes/pass.cpp [hunk 1]
```diff
@@ -293,6 +293,9 @@ void PassRegistry::registerPasses() {
                createMemoryPackingPass);
   registerPass(
     "merge-blocks", "merges blocks to their parents", createMergeBlocksPass);
+  registerPass("merge-data-segments",
+               "merges adjacent active data segments into a single segment",
+               createMergeDataSegmentsPass);
   registerPass("merge-similar-functions",
                "merges similar functions when benefical",
                createMergeSimilarFunctionsPass);
```
Comment (kripken) [inline src/passes/pass.cpp:301]: This pass should added to the default pipeine (unless there is a reason not to). See e.g. where the segment packing pass is added, it could be right before/after.

File: test/lit/passes/merge-data-segments-tnh.wast [hunk 2]
```diff
@@ -0,0 +1,147 @@
+;; NOTE: Assertions have been generated by update_lit_checks.py --all-items and should not be edited.
+;; RUN: foreach %s %t wasm-opt -all --merge-data-segments -tnh -S -o - | filecheck %s
+
+;; Guaranteed traps remain guaranteed under TNH.
```
Comment (kripken) [inline test/lit/passes/merge-data-segments-tnh.wast]: Please expand the comments to save the reader time. E.g. here, something like

```suggestion
;; Guaranteed traps remain guaranteed under TNH. The second segment of memory 1 will trap here, as the memory is empty, so we cannot merge it to the one before.
```

What does the second memory add to this test?

**Review by kripken — COMMENTED** [review]
Thanks for the PR!

I haven't read the code yet, but here are some initial comments elsewhere.

File: test/lit/passes/merge-data-segments-tnh.wast (hunk 2 above)
Comment (kripken) [reply test/lit/passes/merge-data-segments-tnh.wast]: (if the second memory is not needed, it could be removed so the testcase is more focused)

File: src/passes/pass.cpp (hunk 1 above)
Comment (LegionMammal978) [reply src/passes/pass.cpp:301]: Yes, I agree that it would be a good idea to add this as a default, but I wasn't quite sure where to put it in the order. Thematically, it might go right next to MemoryPacking, which is quite early. Practically, it wants to follow SimplifyGlobals, so that all possible immutable-global offsets are resolved, but precede RemoveUnusedModuleElements, since it can remove the last reference to a memory. Would immediately before RemoveUnusedModuleElements be a good place to put it?

Comment (kripken) [reply src/passes/pass.cpp:301]: Yes, that sounds reasonable.

File: test/lit/passes/merge-data-segments-tnh.wast (hunk 2 above)
Comment (LegionMammal978) [reply test/lit/passes/merge-data-segments-tnh.wast]: I've expanded on most of the test comments. I've clarified that "We use a second memory for the dead segment to distinguish between the trap behavior and the drop behavior," i.e., the second memory just adds a bit of extra isolation from what is being done in the first memory. In particular, I want the dead segment to be perfectly valid per se: it only gets dropped since it appears after a trap segment.

File: src/passes/pass.cpp (hunk 1 above)
Comment (LegionMammal978) [reply src/passes/pass.cpp:301]: Alright, I've added it to the list.

**Comment by LegionMammal978** [comment]
After rebasing onto e29edaea576a4ecba41d063b54d0ee5e5223820b and commenting out the `(pagesize 1)` tests, I'm able to get a 12-hour fuzz to succeed with the new pass:
```
Invocations so far:
   FuzzExec: 18711
   CompareVMs: 18863
   CheckDeterminism: 3020
   Wasm2JS: 78
   TrapsNeverHappen: 1717
   CtorEval: 651
   Merge: 1063
   RoundtripText: 764
   ClusterFuzz: 1555
   Two: 18484
   PreserveImportsExportsRandom: 1510
   PreserveImportsExportsJS: 6616
   BranchHintPreservation: 1482
```

File: test/lit/passes/merge-data-segments.wast [hunk 3]
```diff
@@ -0,0 +1,897 @@
...
+  ;; CHECK:      (memory $0 1 1)
+  (memory $0 1 1)
+  ;; CHECK:      (data $0 (i32.const 3) "foobarpezqux")
+  (data $0 (i32.const 3) "foo")
+  (data $1 (i32.const 6) "bar")
+  (data $2 (i32.const 9) "pez")
+  (data $3 (i32.const 12) "qux")
+)
+
+(module
```
Comment (kripken) [inline test/lit/passes/merge-data-segments.wast:28]: These modules can also use comments. They make it a lot easier to go through large PRs like this (and later, if we find a fuzz bug or something, it can help then as well).

For this module, for example, I would  add
```suggestion
;; As above, but with the order of segments reversed. We can still merge them.
(module
```

File: test/lit/passes/merge-data-segments.wast [hunk 4]
```diff
@@ -0,0 +1,897 @@
...
+  ;; CHECK:      (memory $0 1 1)
+  (memory $0 1 1)
+  (data $0 (i32.const 12) "qux")
+  (data $1 (i32.const 9) "pez")
+  (data $2 (i32.const 6) "bar")
+  ;; CHECK:      (data $3 (i32.const 3) "foobarpezqux")
+  (data $3 (i32.const 3) "foo")
+)
+
+(module
```
Comment (kripken) [inline test/lit/passes/merge-data-segments.wast:45]: Another example:
```suggestion
;; Overlapping segments. Here the last tramples those before.
(module
```

File: test/lit/passes/merge-data-segments.wast (hunk 3 above)
Comment (LegionMammal978) [reply test/lit/passes/merge-data-segments.wast:28]: I've elaborated on the comments for all the basic merge cases, giving some reasoning for the name and data of the resulting merged segments. For this module:
```
;; The same merge, but with the segments appearing in reverse order. The
;; resulting data should be the same, since adjacent merge respects address
;; order instead of appearance order. Even though segment $3 appears last, the
;; merged segment takes its name since it has the lowest start address.
```

File: test/lit/passes/merge-data-segments.wast (hunk 4 above)
Comment (LegionMammal978) [reply test/lit/passes/merge-data-segments.wast:45]: I've added a subsection comment explaining the order-of-appearance behavior for overlapping segments:
```
;; The following tests cover overlapping merge cases. Per the spec, active data
;; segments are written to their memories in order of appearance, so when they
;; overlap, we should take the data from the last-appearing segment for each
;; address.
```
As well as another comment for that particular module:
```
;; Growing segments with a common start address. We take the name of segment $0
;; since it is the first-appearing segment at address 0.
```

File: test/lit/passes/merge-data-segments-tnh.wast [hunk 5]
```diff
@@ -0,0 +1,161 @@
+;; NOTE: Assertions have been generated by update_lit_checks.py --all-items and should not be edited.
+;; RUN: foreach %s %t wasm-opt -all --merge-data-segments -tnh -S -o - | filecheck %s
+
+;; Guaranteed traps remain guaranteed under TNH: these are identical to the
```
Comment (kripken) [inline test/lit/passes/merge-data-segments-tnh.wast]: TNH is allowed to ignore guaranteed traps. Is there a benefit to preserving the traps, like it makes the code simpler? Usually it is simpler the other way in my experience but I haven't read the code carefully yet.

File: test/lit/passes/merge-data-segments.wast [hunk 6]
```diff
@@ -0,0 +1,934 @@
...
+      (i32.const 0)
+      (i32.const 0)
+      (i32.const 0)
+    )
+    (memory.init $2
+      (i32.const 0)
+      (i32.const 0)
+      (i32.const 0)
+    )
+    (data.drop $0)
```
Comment (kripken) [inline test/lit/passes/merge-data-segments.wast:311]: Why is it valid for this drop of 0 to turn into a drop of the merged 2?

File: test/lit/passes/merge-data-segments-tnh.wast (hunk 5 above)
Comment (LegionMammal978) [reply test/lit/passes/merge-data-segments-tnh.wast]: There's definitely the idea that "Under TNH, if there's a function/branch that would always trap when executed, we can assume it is never executed, and replace the whole body with `unreachable`." But there's no easy equivalent notion of "If a module can never be instantiated, we can replace the whole module with `unreachable`," except perhaps for throwing a fatal error and refusing to output anything. So for guaranteed traps, I just use the same behavior for TNH and non-TNH.

In principle, in the non-TNH case we could similarly be aggressive in eliminating non-exported function bodies, but I didn't want to complicate the rewriting logic any further for what should hopefully be a rare edge case.

File: test/lit/passes/merge-data-segments.wast (hunk 6 above)
Comment (LegionMammal978) [reply test/lit/passes/merge-data-segments.wast:311]: All of the instructions taking a data-segment index are intended to be used with passive segments. In fact, it's valid in the spec to call them with active segments, but the instantiation process should effectively call `(data.drop $seg)` on each active data segment immediately after using it for initialization, see https://webassembly.github.io/spec/core/exec/modules.html#aux-rundata. Thus, if a module explicitly tries to pass an active segment into any instruction, it will act just like an empty passive segment.

Since every active segment acts exactly the same for these instructions, we just need to rename them all into 'some' active segment: the name and contents don't matter at the point of execution. (In fact, MemoryPacking goes further in this case, and rewrites those instructions with no-ops and pure bounds checks as appropriate. But I wanted to keep the function rewriting simple for this pass.)

File: test/lit/passes/merge-data-segments-tnh.wast (hunk 5 above)
Comment (LegionMammal978) [reply test/lit/passes/merge-data-segments-tnh.wast]: I've added this explanation to the test comments.

File: test/lit/passes/merge-data-segments.wast (hunk 6 above)
Comment (LegionMammal978) [reply test/lit/passes/merge-data-segments.wast:311]: I've added this explanation to the test comments.

File: test/lit/passes/merge-data-segments-tnh.wast [hunk 7]
```diff
@@ -0,0 +1,164 @@
+;; NOTE: Assertions have been generated by update_lit_checks.py --all-items and should not be edited.
+;; RUN: foreach %s %t wasm-opt -all --merge-data-segments -tnh -S -o - | filecheck %s
+
+;; Guaranteed traps remain guaranteed under TNH. It's nontrivial to "drop
+;; absolutely everything" from a module that can never be instantiated, so to
+;; simplify the implementation, we just reuse the non-TNH behavior. Thus, these
+;; are identical to the non-TNH tests for guaranteed traps. We use a second
```
Comment (kripken) [inline test/lit/passes/merge-data-segments-tnh.wast:7]: Ah, if these are just identical to the non-tnh case, then please add another run line (`RUN` in lit) to a single file. That is, the single test file can run in both tnh and non-tnh modes, and verify their output is identical.

Comment (kripken) [reply test/lit/passes/merge-data-segments-tnh.wast:7]: This comment can then be added there to explain why the outputs are identical.
